test/integration/garbagecollector/garbage_collector_test.go

```diff
@@ -1276,3 +1288,91 @@ func testCRDDeletion(t *testing.T, ctx *testContext, ns *v1.Namespace, definitio
 		t.Fatalf("failed waiting for dependent %q (owned by %q) to be deleted", dependent.GetName(), owner.GetName())
 	}
 }
+
+func TestDiscoveryFailedOnGCStart(t *testing.T) {
+	ctx := setup(t, 0)
+	defer ctx.tearDown()
+	apiRegClient, clientSet := ctx.apiRegistrationClient, ctx.clientSet
+
+	// Create a fake apiservice to simulate discovery failed
+	apiservice, err := apiRegClient.APIServices().Create(context.TODO(), &apiregistration.APIService{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "APIService",
+			APIVersion: "apiregistration.k8s.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "v1.test.io",
+		},
+		Spec: apiregistration.APIServiceSpec{
+			Service: &apiregistration.ServiceReference{
+				Namespace: "default",
+				Name:      "test",
+			},
+			Group:                "test.io",
+			Version:              "v1",
+			GroupPriorityMinimum: 1,
+			VersionPriority:      1,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create apiservice: %v", err)
+	}
+	defer apiRegClient.APIServices().Delete(context.TODO(), apiservice.Name, metav1.DeleteOptions{})
+
+	ns := createNamespaceOrDie("gc-discovery-fail", clientSet, t)
+	defer deleteNamespaceOrDie(ns.Name, clientSet, t)
+
+	testBlockingOwnerRefDoesBlock(t, ctx, ns, 1*time.Second)
```

Comment:
why do we have to inject a one second delay here (what are we waiting for)?

rather than plumbing delays into test methods and helpers, can we poll for whatever it is we're waiting for before calling testBlockingOwnerRefDoesBlock instead?

Reply:
> why do we have to inject a one second delay here (what are we waiting for)?

Without the delay, the `gc.Run` will run with all resource monitors have been synced, due to the `gc.Sync` have not add any resource monitors yet.

And we want to test the case that `gc.Run` with all resource monitors not have been synced.